src/librustc/mir/interpret/queries.rs

```diff
@@ -76,3 +78,31 @@ impl<'tcx> TyCtxt<'tcx> {
         self.const_eval_validated(param_env.and(cid))
     }
 }
+
+impl<'infcx, 'tcx> InferCtxt<'infcx, 'tcx> {
+    /// Resolves and evaluates a constant.
+    ///
+    /// The constant can be located on a trait like `<A as B>::C`, in which case the given
+    /// substitutions and environment are used to resolve the constant. Alternatively if the
+    /// constant has generic parameters in scope the substitutions are used to evaluate the value of
+    /// the constant. For example in `fn foo<T>() { let _ = [0; bar::<T>()]; }` the repeat count
+    /// constant `bar::<T>()` requires a substitution for `T`, if the substitution for `T` is still
+    /// too generic for the constant to be evaluated then `Err(ErrorHandled::TooGeneric)` is
+    /// returned.
+    ///
+    /// This handles inferences variables within both `param_env` and `substs` by
+    /// canonicalizing both values before resolving and evaluating the constant.
+    pub fn const_eval_resolve(
+        &self,
+        param_env: ty::ParamEnv<'tcx>,
+        def_id: DefId,
+        substs: SubstsRef<'tcx>,
+        promoted: Option<mir::Promoted>,
+        span: Option<Span>,
+    ) -> ConstEvalResult<'tcx> {
+        let mut query_state = OriginalQueryValues::default();
+        let canonical = self.canonicalize_query(&(param_env, substs), &mut query_state);
+        let (param_env, substs) = canonical.value;
+        self.tcx.const_eval_resolve(param_env, def_id, substs, promoted, span)
```

Comment:
Wait, I just remembered something: you need to replace the type with the original one, *if* you want to return the value (since it's not used, you could just opt to return `Result<(), ...>` for now).
This is because you don't want the canonical variables to stick around in the type.

There's also a related trick I've told @nikomatsakis about before:
The *value* produced by const-eval shouldn't depend on lifetimes, so we should probably always erase them in the input `substs` (and `param_env` even?).
But the type might have relevant lifetimes, which means we should (just like in the canonical case) use the original type along the new value.

I believe both canonicalizing and erasing lifetimes, would produce the correct value, but erasing them would result in perhaps better cache utilization.

Reply:
Ideally, canonicalization would replace *all* lifetimes and hence be the same, from a cache perspective, but we may not currently handle things that way (we may treat `'static` as special or something...)

Reply:
I would think erasing lifetimes would also be benefical for const eval queries which don't use canonicalization, i.e. the ones that don't use `InferCtxt::const_eval_reolve`, although in most cases they already probably have there regions erased. It does look like `canonicalize_query`, which i'm using, does replace `static lifetimes.

Reply:
> But the type might have relevant lifetimes, which means we should (just like in the canonical case) use the original type along the new value.

Most usage of const-eval is by codegen or inside miri which doesn't need/want lifetimes, and in other cases the caller knows what type it's expecting, so I think it should be up to the callers to the replace the type if they need to. One place I'm unsure about is: https://github.com/skinny121/rust/blob/c8e35e3df4016e72bf7eb3a1508a592a80d1f3ae/src/librustc_mir_build/hair/pattern/mod.rs#L764

I'm not sure if lifetimes are important here or not, if they are I can replace the type with the node type retrieved above.